azure/datalake/store/lib.py

```diff
@@ -136,60 +131,63 @@ def auth(tenant_id=None, username=None,
 
     if not client_secret:
         client_secret = os.environ.get('azure_client_secret', None)
+        contextPub = msal.PublicClientApplication(client_id=client_id, authority=authority+tenant_id)
```

Comment:
This PR is currently a 1:1 translation from ADAL to MSAL. It would work no worse than before, but it also does not utilize many good features from MSAL, such as token cache.

Ideally we shall reuse MSAL's `PublicClientApplication` and `ConfidentialClientApplication` (as mentioned in the first Q&A of [this recent conversation](https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/549)), but that would need a major architecture change to this repo. We discussed offline that this repo has a shelf life for less than a year, and [its replacement](https://learn.microsoft.com/en-us/samples/azure/azure-sdk-for-python/storage-datalake-samples/) is already available, so, we choose to use the current "1:1 translation" approach.

There is one low-hanging fruit we can grab. Define a module-wide in-memory dictionary, for MSAL to use it as an [http cache](https://msal-python.readthedocs.io/en/latest/#msal.PublicClientApplication.params.http_cache). This is required for MSAL's built-in http 429/5xx error retry to work across multiple `lib.auth(...)` runs.

```suggestion
        contextPub = msal.PublicClientApplication(client_id=client_id, authority=authority+tenant_id, http_cache=_http_cache)
```